src/ctypes/ldouble_stubs.c

```diff
@@ -0,0 +1,525 @@
+  int i = Int_val(vi);
+  long double rf = ldexpl(f, i);
+  r = ctypes_copy_ldouble(rf);
+  CAMLreturn(r);
+}
+
+CAMLprim value ctypes_ldouble_modf(value v) {
+  CAMLparam1(v);
+  CAMLlocal1(r);
+  long double f = ldouble_custom_val(v);
+  long double rf2;
+  long double rf1 = modfl(f, &rf2);
+  r = caml_alloc_tuple(2);
+  Store_field(r, 0, ctypes_copy_ldouble(rf1));
+  Store_field(r, 1, ctypes_copy_ldouble(rf2));
+  CAMLreturn(r);
+}
+
+enum {
+  ml_FP_NORMAL = 0,
+  ml_FP_SUBNORMAL,
+  ml_FP_ZERO,
+  ml_FP_INFINITE,
+  ml_FP_NAN,
+};
+
+CAMLprim value ctypes_ldouble_classify(value v){
+  CAMLparam1(v);
+  CAMLlocal1(r);
+  long double f = ldouble_custom_val(v);
+  switch (fpclassify(f)){
+  case FP_NORMAL    : r = Val_int(ml_FP_NORMAL); break;
+  case FP_SUBNORMAL : r = Val_int(ml_FP_SUBNORMAL); break;
+  case FP_ZERO      : r = Val_int(ml_FP_ZERO); break;
+  case FP_INFINITE  : r = Val_int(ml_FP_INFINITE); break;
+  case FP_NAN       : 
+  default           : r = Val_int(ml_FP_NAN); break;
+  }
+  CAMLreturn(r);
+}
+
+static char *format_ldouble(char *fmt, long double d) {
+  static size_t buf_len = 10;
+  static char *buf = NULL;
+  static char *empty = "";
+  size_t print_len = 0;
+
+  // allocate initial string
+  if (buf == NULL) buf = malloc(buf_len);
+  if (buf == NULL) return empty; // oops...stuck
+
+  // try to print
+  print_len = snprintf(buf, buf_len, fmt, d);
+
+  // not enough space - reallocate and try again
+  if (print_len >= buf_len) {
+    // re-allocate string
+    if (buf) free(buf);
+    buf = malloc(print_len+1);
+    buf_len = print_len+1;
+    // try again
+    return format_ldouble(fmt, d);
+  } else 
+    // ok
+    return buf;
+}
+
+CAMLprim value ctypes_ldouble_format(value fmt, value d) {
+  CAMLparam2(fmt, d);
+  CAMLreturn(caml_copy_string(format_ldouble( String_val(fmt), ldouble_custom_val(d))));
+}
+
+CAMLprim value ctypes_ldouble_of_string(value v) {
+  CAMLparam1(v);
+  char *str = String_val(v);
+  char *end = str + caml_string_length(v);
+  long double r = strtold(str, &end);
+  CAMLreturn(ctypes_copy_ldouble(r));
+}
+
+char hex_char(char x) {
+  if (x < 10) return '0' + x;
+  return 'a' + x - 10;
+}
+
+CAMLprim value ctypes_ldouble_to_hex(value v) {
+  CAMLparam1(v);
+  static char x[LDOUBLE_STORAGE_BYTES*2 + 1];
+  char *p = (char *) Data_custom_val(v);
+  int i;
+  for (i=0; i<LDOUBLE_STORAGE_BYTES; i++) {
+    x[i*2+0] = hex_char(((*(p+i)) >> 0) & 0xf);
+    x[i*2+1] = hex_char(((*(p+i)) >> 4) & 0xf);
+  }
+  x[LDOUBLE_STORAGE_BYTES*2] = 0;
+  CAMLreturn(caml_copy_string(x));
+}
+
+value ctypes_ldouble_min(value unit) { return ctypes_copy_ldouble(-LDBL_MAX); }
+value ctypes_ldouble_max(value unit) { return ctypes_copy_ldouble(LDBL_MAX); }
+value ctypes_ldouble_epsilon(value unit) { return ctypes_copy_ldouble(LDBL_EPSILON); }
+value ctypes_ldouble_nan(value unit) { return ctypes_copy_ldouble(nanl("char-sequence")); }
+// XXX note; -(log 0) gives +ve inf (and vice versa).  Is this consistent? *)
```

Comment:
I was more concerned that different c-libraries would ensure the same behaviour - I think I have convinced myself that they mathematically _should_ but it's maybe worth keeping the comment just in case.